File: budgetadvisor-ai/backend/app/api/routes/analytics.py
"""
Analytics endpoints — user-scoped, derived from receipts and items.
"""

from collections import defaultdict
from datetime import date, timedelta
from typing import Any, Literal

from fastapi import APIRouter
from sqlmodel import col, select

from app.api.deps import CurrentUser, SessionDep
from app.models import Receipt, ReceiptItem
from app.models.analytics import (
    CategoryBreakdownItem,
    DashboardStats,
    TopMerchant,
    WeeklyTrendItem,
)

router = APIRouter(prefix="/analytics", tags=["analytics"])

# Category display metadata — matches frontend CATEGORIES constant
CATEGORY_META: dict[str, tuple[str, str]] = {
    "dairy": ("Dairy", "#6366F1"),
    "meat": ("Meat", "#F43F5E"),
    "fresh_produce": ("Fresh Produce", "#22C55E"),
    "bakery": ("Bakery", "#F59E0B"),
    "processed_foods": ("Processed Foods", "#8B5CF6"),
    "household": ("Household", "#06B6D4"),
    "personal_care": ("Personal Care", "#EC4899"),
    "beverages": ("Beverages", "#14B8A6"),
    "other": ("Other", "#94A3B8"),
}

Period = Literal["week", "month", "last_month"]


def _period_start(period: Period) -> date:
    today = date.today()
    if period == "week":
        return today - timedelta(days=today.weekday())
    if period == "last_month":
        first_of_this = today.replace(day=1)
        return (first_of_this - timedelta(days=1)).replace(day=1)
    return today.replace(day=1)  # month (default)


def _period_end(period: Period) -> date:
    today = date.today()
    if period == "last_month":
        return today.replace(day=1) - timedelta(days=1)
    return today


def _get_period_receipts(
    session: SessionDep, user_id: Any, period: Period
) -> list[Receipt]:
    start = _period_start(period)
    end = _period_end(period)
    return session.exec(
        select(Receipt)
        .where(Receipt.user_id == user_id)
        .where(col(Receipt.transaction_date) >= start)
        .where(col(Receipt.transaction_date) <= end)
    ).all()


# ---------------------------------------------------------------------------
# GET /analytics/dashboard
# ---------------------------------------------------------------------------


@router.get("/dashboard", response_model=DashboardStats)
def get_dashboard_stats(
    current_user: CurrentUser,
    session: SessionDep,
    period: Period = "month",
) -> Any:
    receipts = _get_period_receipts(session, current_user.id, period)

    this_month = round(sum(r.total_amount or 0 for r in receipts), 2)
    saved = round(sum(r.total_saved for r in receipts), 2)
    budget = current_user.monthly_budget
    budget_percent = round((this_month / budget * 100), 1) if budget else 0.0

    return DashboardStats(
        this_month=this_month,
        budget=budget,
        budget_percent=budget_percent,
        saved=saved,
        receipt_count=len(receipts),
    )


# ---------------------------------------------------------------------------
# GET /analytics/categories
# ---------------------------------------------------------------------------


@router.get("/categories", response_model=list[CategoryBreakdownItem])
def get_category_breakdown(
    current_user: CurrentUser,
    session: SessionDep,
    period: Period = "month",
) -> Any:
    receipts = _get_period_receipts(session, current_user.id, period)
    if not receipts:
        return []

    receipt_ids = [r.id for r in receipts]
    items = session.exec(
        select(ReceiptItem).where(col(ReceiptItem.receipt_id).in_(receipt_ids))
    ).all()

    totals: dict[str, float] = defaultdict(float)
    for item in items:
        totals[item.category] += item.line_total or 0

    grand_total = sum(totals.values()) or 1

    return sorted(
        [
            CategoryBreakdownItem(
                key=key,
                label=CATEGORY_META.get(
                    key, (key.replace("_", " ").title(), "#94A3B8")
                )[0],
                amount=round(amount, 2),
                color=CATEGORY_META.get(key, (key, "#94A3B8"))[1],
                percent=round((amount / grand_total) * 100, 1),
            )
            for key, amount in totals.items()
            if amount > 0
        ],
        key=lambda c: c.amount,
        reverse=True,
    )


# ---------------------------------------------------------------------------
# GET /analytics/weekly-trend
# ---------------------------------------------------------------------------


@router.get("/weekly-trend", response_model=list[WeeklyTrendItem])
def get_weekly_trend(
    current_user: CurrentUser,
    session: SessionDep,
) -> Any:
    eight_weeks_ago = date.today() - timedelta(weeks=8)
    receipts = session.exec(
        select(Receipt)
        .where(Receipt.user_id == current_user.id)
        .where(col(Receipt.transaction_date) >= eight_weeks_ago)
    ).all()

    # Accumulate per ISO-week key, preserving insertion order
    week_totals: dict[str, float] = defaultdict(float)
    for r in receipts:
        if r.transaction_date:
            key = f"W{r.transaction_date.isocalendar()[1]}"
            week_totals[key] += r.total_amount or 0

    # Build last-8-weeks ordered list
    today = date.today()
    seen: set[str] = set()
    result: list[WeeklyTrendItem] = []
    for i in range(7, -1, -1):
        d = today - timedelta(weeks=i)
        label = f"W{d.isocalendar()[1]}"
        if label not in seen:
            seen.add(label)
            result.append(
                WeeklyTrendItem(week=label, amount=round(week_totals.get(label, 0), 2))
            )

    return result


# ---------------------------------------------------------------------------
# GET /analytics/top-merchants
# ---------------------------------------------------------------------------


@router.get("/top-merchants", response_model=list[TopMerchant])
def get_top_merchants(
    current_user: CurrentUser,
    session: SessionDep,
    period: Period = "month",
) -> Any:
    receipts = _get_period_receipts(session, current_user.id, period)

    merchant_data: dict[str, dict[str, Any]] = defaultdict(
        lambda: {"visits": 0, "total": 0.0}
    )
    for r in receipts:
        name = r.merchant_name_raw or "Unknown"
        merchant_data[name]["visits"] += 1
        merchant_data[name]["total"] += r.total_amount or 0

    return sorted(
        [
            TopMerchant(
                name=name,
                visits=d["visits"],
                total=round(d["total"], 2),
                avg=round(d["total"] / d["visits"], 2),
            )
            for name, d in merchant_data.items()
        ],
        key=lambda m: m.total,
        reverse=True,
    )[:10]

File: budgetadvisor-ai/frontend/src/lib/analytics.ts
/**
 * Pure functions that derive analytics from ReceiptWithItems[] data.
 * All computed client-side from receipts already loaded — no extra API calls.
 */
import type {
  CategoryBreakdownItem,
  DashboardStats,
  TopMerchant,
  WeeklyTrendItem,
  ReceiptWithItems,
} from "@/api/types";

export type AnalyticsPeriod = "week" | "month" | "last_month";

// ---------------------------------------------------------------------------
// Category display metadata — matches CATEGORIES in dummy-data.ts
// ---------------------------------------------------------------------------
const CATEGORY_META: Record<string, [string, string]> = {
  dairy:           ["Dairy",           "#6366F1"],
  meat:            ["Meat",            "#F43F5E"],
  fresh_produce:   ["Fresh Produce",   "#22C55E"],
  bakery:          ["Bakery",          "#F59E0B"],
  processed_foods: ["Processed Foods", "#8B5CF6"],
  household:       ["Household",       "#06B6D4"],
  personal_care:   ["Personal Care",   "#EC4899"],
  beverages:       ["Beverages",       "#14B8A6"],
  other:           ["Other",           "#94A3B8"],
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function round(n: number, dp = 2): number {
  return Math.round(n * 10 ** dp) / 10 ** dp;
}

/**
 * Parse "2024-03-15" or "2024-03-15T10:30:00Z" → local midnight Date.
 * IMPORTANT: `new Date("YYYY-MM-DD")` is UTC midnight per spec, which causes
 * off-by-one errors when compared against local-midnight period bounds.
 * Parsing components manually creates a local date instead.
 */
function toDateOnly(s: string): Date {
  const [y, m, d] = s.split("T")[0].split("-").map(Number);
  return new Date(y, m - 1, d); // local midnight — safe to compare with getPeriodBounds()
}

/** ISO week number */
function isoWeek(d: Date): number {
  const utc = new Date(Date.UTC(d.getFullYear(), d.getMonth(), d.getDate()));
  const day = utc.getUTCDay() || 7;
  utc.setUTCDate(utc.getUTCDate() + 4 - day);
  const yearStart = new Date(Date.UTC(utc.getUTCFullYear(), 0, 1));
  return Math.ceil(((utc.getTime() - yearStart.getTime()) / 86_400_000 + 1) / 7);
}

/** Get the best available date for a receipt (transaction_date preferred) */
function receiptDate(r: ReceiptWithItems): Date | null {
  const s = r.transaction_date ?? r.created_at;
  return s ? toDateOnly(s) : null;
}

// ---------------------------------------------------------------------------
// Period filtering
// ---------------------------------------------------------------------------

export function getPeriodBounds(period: AnalyticsPeriod): [Date, Date] {
  const now = new Date();
  const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());

  if (period === "week") {
    const start = new Date(today);
    start.setDate(today.getDate() - ((today.getDay() + 6) % 7)); // Monday
    return [start, today];
  }
  if (period === "last_month") {
    const end = new Date(now.getFullYear(), now.getMonth(), 0); // last day prev month
    const start = new Date(end.getFullYear(), end.getMonth(), 1);
    return [start, end];
  }
  // month (default)
  return [new Date(now.getFullYear(), now.getMonth(), 1), today];
}

export function filterByPeriod(
  receipts: ReceiptWithItems[],
  period: AnalyticsPeriod
): ReceiptWithItems[] {
  const [start, end] = getPeriodBounds(period);
  return receipts.filter((r) => {
    const d = receiptDate(r);
    if (!d) return false;
    return d >= start && d <= end;
  });
}

// ---------------------------------------------------------------------------
// Stat computations
// ---------------------------------------------------------------------------

export function computeDashboardStats(
  receipts: ReceiptWithItems[],
  monthlyBudget: number | null
): DashboardStats {
  const thisMonth = receipts.reduce((s, r) => s + (r.total_amount ?? 0), 0);
  const saved = receipts.reduce((s, r) => s + r.total_saved, 0);
  const budgetPercent =
    monthlyBudget && monthlyBudget > 0
      ? round((thisMonth / monthlyBudget) * 100, 1)
      : 0;
  return {
    this_month: round(thisMonth),
    budget: monthlyBudget,
    budget_percent: budgetPercent,
    saved: round(saved),
    receipt_count: receipts.length,
  };
}

export function computeCategoryBreakdown(
  receipts: ReceiptWithItems[]
): CategoryBreakdownItem[] {
  const totals: Record<string, number> = {};
  for (const r of receipts) {
    for (const item of r.items) {
      totals[item.category] = (totals[item.category] ?? 0) + (item.line_total ?? 0);
    }
  }
  const grand = Object.values(totals).reduce((s, v) => s + v, 0) || 1;
  return Object.entries(totals)
    .filter(([, v]) => v > 0)
    .map(([key, amount]) => ({
      key,
      label: CATEGORY_META[key]?.[0] ?? key.replace(/_/g, " "),
      amount: round(amount),
      color: CATEGORY_META[key]?.[1] ?? "#94A3B8",
      percent: round((amount / grand) * 100, 1),
    }))
    .sort((a, b) => b.amount - a.amount);
}

export function computeWeeklyTrend(receipts: ReceiptWithItems[]): WeeklyTrendItem[] {
  const weekTotals: Record<string, number> = {};
  for (const r of receipts) {
    const d = receiptDate(r);
    if (!d) continue;
    const label = `W${isoWeek(d)}`;
    weekTotals[label] = (weekTotals[label] ?? 0) + (r.total_amount ?? 0);
  }

  // Build last-8-weeks ordered list (no duplicates)
  const today = new Date();
  const result: WeeklyTrendItem[] = [];
  const seen = new Set<string>();
  for (let i = 7; i >= 0; i--) {
    const d = new Date(today);
    d.setDate(today.getDate() - i * 7);
    const label = `W${isoWeek(d)}`;
    if (!seen.has(label)) {
      seen.add(label);
      result.push({ week: label, amount: round(weekTotals[label] ?? 0) });
    }
  }
  return result;
}

export function computeTopMerchants(receipts: ReceiptWithItems[]): TopMerchant[] {
  const data: Record<string, { visits: number; total: number }> = {};
  for (const r of receipts) {
    const name = r.merchant_name_raw ?? "Unknown";
    if (!data[name]) data[name] = { visits: 0, total: 0 };
    data[name].visits += 1;
    data[name].total += r.total_amount ?? 0;
  }
  return Object.entries(data)
    .map(([name, d]) => ({
      name,
      visits: d.visits,
      total: round(d.total),
      avg: round(d.total / d.visits),
    }))
    .sort((a, b) => b.total - a.total)
    .slice(0, 10);
}
